src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java

```diff
@@ -0,0 +1,390 @@
+    * Same as calling {@link #kill()} then {@link #join()}.
+    * InterruptedExceptions are ignored. To handle interrupted exceptions,
+    * call {@link #kill()} then {@link #join()} manually.
+    */
+   public void blockingKill()
+   {
+      kill();
+      try
+      {
+         join();
+      }
+      catch (InterruptedException ignored) {}
+   }
+
+   /**
+    * The method that is executed repeatedly in a loop.
+    * <p>
+    * You may {@code @Override} this method with the code to repeat.
+    * Alternatively, if a {@link RunnableThatThrows} was passed it, this method
+    * will call the run method in the loop.
+    *
+    * @throws Throwable Any throwable that the overriding code or the passed in task may throw.
+    *       This throwable will be handled by the passed in {@link ExceptionHandler}
+    *       (by default it is {@link DefaultExceptionHandler#MESSAGE_AND_STACKTRACE}).
+    */
+   protected void repeat() throws Throwable
+   {
+      if (task != null)
+         task.run();
+   }
+
+   /**
+    * The {@link Thread#run()} method, overridden to run a loop.
+    * To extend this class {@link Thread} style, override {@link #repeat()} instead.
+    * <p>
+    * DO NOT CALL THIS METHOD. Well, you can, but why would you?
+    * You are using a thread to run things asynchronously, but calling this would run the loop synchronously.
+    * Why would you want that?
+    * <p>
+    * This method is a necessary evil committed for this class to extend Thread.
+    */
+   @Override
+   public final void run()
+   {
+      while (running)
+      {
+         try
+         {
+            synchronized (loopLock)
+            {  // No more runs remaining -> wait until something changes
+               if (remainingRepetitions == 0)
+               {
+                  loopLock.wait();
+                  continue;
+               }
+
+               // Decrement the counter for the run that's about to occur
+               if (remainingRepetitions > 0)
+                  remainingRepetitions--;
+            }
+
+            // If a period/frequency limit was set, wait until loop can run.
+            if (loopPeriodLowerLimit > 0.0)
+            {
+               /*
+                * This call must not swallow interrupts.
+                * As of writing this comment (Nov, 2024), LockSupport.parkNanos() is used internally to block.
+                * Although the throttler will block until the period has elapsed, the thread
+                * remains interrupted.
+                */
+               throttler.waitAndRun(loopPeriodLowerLimit);
+            }
+         }
+         catch (InterruptedException interrupted)
+         {  // Maintain interrupted status so that runInLoop can handle it
+            interrupt();
+         }
+
+         // Run the runInLoop method, and handle any exception it may throw.
+         ExceptionTools.handle(this::repeat, exceptionHandler);
+         ++completedRepetitions;
```

Comment:
What if you decrement remaining down here, so then you don't have a moment where remaining == 0 but it's working on executing the task.

Reply:
When the task is being executed the repetition has begun, so I'd say 0 remain to be executed. At least that's how I've been thinking about it. Also it's nice to keep it in one `synchronized` block haha. I'm still willing to make this change if you insist

Reply:
Currently there is no way to tell if a task is currently executing or not. It would be great is repetitions have 3 states SCHEDULED -> EXECUTING -> COMPLETED and that be thread safe state. I would advocate adding a boolean for isExecuting, but also, I think the "remaining" term is ambiguous for whether or not it's remaining _while_ it's executing.